pkg/plugins/policies/xds/meshroute/listeners.go

```diff
@@ -42,3 +48,72 @@ func GetClusterName(
 
 	return name
 }
+
+func MakeSplit(
+	proxy *core_xds.Proxy,
+	clusterCache map[string]string,
+	sc *SplitCounter,
+	servicesAcc envoy_common.ServicesAccumulator,
+	refs []common_api.BackendRef,
+) []*plugins_xds.Split {
+	var split []*plugins_xds.Split
+	clusterBuilders := map[string]*plugins_xds.ClusterBuilder{}
+	weights := map[string]uint32{}
+
+	for _, ref := range refs {
+		switch ref.Kind {
+		case common_api.MeshService, common_api.MeshServiceSubset:
+		default:
+			continue
+		}
+
+		service := ref.Name
+		if pointer.DerefOr(ref.Weight, 1) == 0 {
+			continue
+		}
+
+		clusterName := GetClusterName(ref.Name, ref.Tags, sc)
+		isExternalService := plugins_xds.HasExternalService(proxy.Routing, service)
+		refHash := ref.TargetRef.Hash()
+		refWeight := uint32(pointer.DerefOr(ref.Weight, 1))
+		weights[refHash] += refWeight
+
+		if existingClusterName, ok := clusterCache[refHash]; ok {
+			// cluster already exists, so adding only split
+			split = append(split, plugins_xds.NewSplitBuilder().
+				WithClusterName(existingClusterName).
+				WithWeight(refWeight).
+				WithExternalService(isExternalService).
+				Build())
+			continue
+		}
+
+		clusterCache[refHash] = clusterName
+
+		split = append(split, plugins_xds.NewSplitBuilder().
+			WithClusterName(clusterName).
+			WithWeight(refWeight).
+			WithExternalService(isExternalService).
+			Build())
+
+		clusterBuilder := plugins_xds.NewClusterBuilder().
+			WithService(service).
+			WithName(clusterName).
+			WithTags(envoy_tags.Tags(ref.Tags).
+				WithTags(mesh_proto.ServiceTag, ref.Name).
+				WithoutTags(mesh_proto.MeshTag)).
+			WithExternalService(isExternalService)
+
+		if mesh, ok := ref.Tags[mesh_proto.MeshTag]; ok {
+			clusterBuilder.WithMesh(mesh)
+		}
+
+		clusterBuilders[refHash] = clusterBuilder
+	}
+
+	for refHash, clusterBuilder := range clusterBuilders {
+		servicesAcc.Add(clusterBuilder.WithWeight(weights[refHash]).Build())
```

Comment:
Why do we need `weight` for the cluster? I think inside Envoy cluster doesn't have weight, moreover, one cluster could be referenced by different listeners with different weight

Reply:
`weight` is currently used only in `TcpProxyConfigurer`, so you still can use the same cluster with different weight whenever it suits you

Reply:
That was one of the reasons to deprecate old [ClusterImpl](https://github.com/kumahq/kuma/blob/master/pkg/xds/envoy/types.go#L28) and split things into 2 entities – Cluster and Split. 

Why can't we use Split in TcpProxyConfigurer?

Reply:
@lobkovilya I'm discussing with @michaelbeaumont the same issue here: https://github.com/kumahq/kuma/pull/6910#discussion_r1214085964 (I tried to explain it there)